parser_library/src/workspaces/processor_group.cpp

```diff
@@ -114,37 +113,32 @@ std::vector<std::pair<std::string, size_t>> processor_group::suggest(std::string
 
 bool processor_group::refresh_needed(const std::vector<utils::resource::resource_location>& urls) const
 {
-    return std::any_of(urls.begin(), urls.end(), [this](const auto& res_location) {
-        constexpr auto matching_prefix = [](std::string_view l, std::string_view r) {
-            const auto common_len = std::min(l.size(), r.size());
-            return l.substr(0, common_len) == r.substr(0, common_len);
-        };
-        const auto has_cached_content = [this](std::span<const size_t> idx) {
-            return std::any_of(idx.begin(), idx.end(), [this](auto i) { return m_libs[i]->has_cached_content(); });
-        };
-
-        const std::string_view url = res_location.get_uri();
-        const auto candidate = m_refresh_prefix.lower_bound(url);
-        return candidate != m_refresh_prefix.end() && matching_prefix(url, candidate->first)
-            && has_cached_content(candidate->second)
-            || candidate != m_refresh_prefix.begin() && matching_prefix(url, std::prev(candidate)->first)
-            && has_cached_content(std::prev(candidate)->second);
+    return std::any_of(urls.begin(), urls.end(), [&libs = m_libs](const auto& res_location) {
```

Comment:
also can the loops switched and has_cached_content tested before entering the nested loop?

Reply:
It's definitely worse than the previous solution, performance-wise. I'll try to think of something else.

One approach, I'm thinking about would be to use `unordered_map<std::string, std::vector<size_t>>` as before (or [trie](https://en.wikipedia.org/wiki/Trie)) but the key would be made of "path segments" in the following way:
- Library with path "test://workspace/externals/library1/SUB_LIBRARY/" would generate several entries and assigns its id to each of them:
"test://workspace/externals/library1/SUB_LIBRARY/", 
"test://workspace/externals/library1/", 
"test://workspace/externals/", 
"test://workspace/"

Then when the `refresh_needed()` is called it would test for each provided `url` if any of its path segments exist in the map. I'm unsure, however, if this would end up being faster.

Reply:
I was also wondering about reducing the size of urls, typically when the vector is bigger than one, it contains a lot of files from the same folder.